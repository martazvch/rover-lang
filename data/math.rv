struct Vec2 {
    x: int,
    y: int,

    fn len(self) -> int {
        self.x + self.y
    }
}

fn introduce() {
    print "This is a mathematical module"
}

fn getVecFromPos(x: int, y: int) -> Vec2 {
    Vec2{ x, y }
}

fn getInt() -> int { 12 }

fn getIntDef(a = 5) -> int { 12 }


