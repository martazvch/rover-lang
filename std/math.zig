pub fn cos(x: f64) f64 {
    return @cos(x);
}

pub fn sin(x: f64) f64 {
    return @sin(x);
}
