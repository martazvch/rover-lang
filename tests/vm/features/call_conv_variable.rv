// All assignment should resolve to the same type and calling convention
// should be transmitted
use data.math 

struct Foo {
    value = 8

    fn static() -> int { 3 }

    fn method(self) -> int {
        self.value
    }

    fn importFuncFromMethod(self) -> fn() -> int {
        math.getInt
    }
}

fn copy() -> int {
    5
}

fn main() {
    var foo = Foo{}

    var bound = Foo.static
    print bound()

    bound = copy
    print bound()

    bound = foo.method
    print bound()

    bound = math.getInt
    print bound()

    // Call conv survives a layer
    var tmp = foo.importFuncFromMethod
    print tmp()
    bound = tmp()
    print bound()
}
