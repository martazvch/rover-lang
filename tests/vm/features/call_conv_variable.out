3
5
8
12
<function getInt>
12
